### ISSUE TITLE
Passing options/state

### ISSUE THREAD

**Issue by Odaeus** [nsweeting/rabbit#64]
Hi,

Thank you for making this! It's well designed and looks great to use.

I've come across one problem in my use-case, that I wish to start consumers dynamically with different configurations. Such as the queue and routing key, or how they behave in reaction to messages. Am I right in saying there's no way to do this as:
- the init options are not provided to `handle_setup`.
- no options or local state are provided to `handle_message`.

I wanted to ask if I'm missing something?

**Comment by nsweeting** [comment]
Hi @Odaeus 

Thanks for the great feedback. To answer your questions:

1. Correct - at the moment, only the channel and queue are passed to `handle_setup/2`. I would actually be open to passing more - a little thought might be needed on this though. Should it simply be the consumer state? https://github.com/nsweeting/rabbit/blob/master/lib/rabbit/consumer/server.ex#L155-L170 Or a subset of it?

2. Correct - at the moment, no custom state/options can be passed along to `handle_message/1`. But I am in 100% agreeance that this should be added as I also have use cases for this that are currently being handled in less elegant ways. The custom options/state would then be made available in the `Rabbit.Message` struct.

Happy to review any PR's submitted for these items.

**Comment by nsweeting** [comment]
Point 2 has been addressed in PR #65. I will look at merging and releasing anew version tomorrow.

**Comment by Odaeus** [comment]
That's great about `handle_message`, thanks!

It is tricky as to whether it should be the state or not as it exposes a lot of internal details. I think in either case, `start_link` should accept arguments, add them to the `state` under e.g. `setup_opts`, and then that or the entire state is passed to `handle_setup`. On balance I'd probably err on the side of flexibility and pass anything that could be relevant from the `state`, or all of it and only document the important keys 😄 .

*Ideally* I'd write something like:

```elixir
DynamicSupervisor.start_child({MySupervisor, [exchange: "foo", routing_key: "bar"]})

...
  def handle_setup(channel, %{setup_opts: setup_opts}) do
    {:ok, queue} = AMQP.Queue.declare(channel) # I'm happy to let RabbitMQ name my queue
    :ok = AMQP.Queue.bind(channel, queue, setup_opts[:exchange], routing_key: setup_opts[:routing_key])
  end
```

As you can see, I don't want to set the `queue` argument, but that's a separate issue entirely!

Let me know what you'd prefer and I believe I'll have time to work on it this week.

**Comment by nsweeting** [comment]
Great - thanks for the example @Odaeus.

Yeah -  I think having a defined key that can be passed in the consumer options (`setup_opts`?)   And maybe we just update `handle_setup/2` to become `handle_setup/3`.

```elixir
Rabbit.Consumer.start_link(MyConsumer, queue: "", setup_opts: [exchange: "", routing_key: "")

...

def handle_setup(channel, queue, options) do
  ...
end
```

Thoughts?

**Comment by Odaeus** [comment]
I started fiddling around and unfortunately went in the opposite direction 😬  of a `handle_setup(state)` that also updates the `state`. As with automatically-named queues the name needs to be persisted otherwise `Consumer.Server.consume` can't start (the weird/undocumented return of `AMQP.Queue.declare` got me!). Obviously there's other options, `handle_setup` could return just the queue name for example.

You can see my current changes here: https://github.com/nsweeting/rabbit/compare/master...Odaeus:handle_setup_state

Adding your proposed third arg to `handle_setup` has the advantage of being clearer and won't break anything if the user returns an invalid state. So I'm not sure which is better.

**Comment by nsweeting** [comment]
You're right - I didnt think of the fact that automatically named queues would need the queue name persisted in the server state. All of my use cases up until now have worked with a strongly defined topology and use the `Rabbit.Broker` `:topology` to set things up before.

I kind of like the idea of only having `handle_setup/1` - passing the state - and having to return the state. It seems clean and along the lines of a typical server callback. If we return the entire server state - there is obviously room for users to shoot themselves in the foot - but I think I'm OK with that as long as we document that fact.

Any documentation that is updated would likely want to highlight the more important keys available in the state (`:channel`, `:queue`, `:setup_opts`).

Also `Rabbit.Producer` has a similar `handle_setup/1` callback that we should probably align with this change.

**Comment by Odaeus** [comment]
Ok great! Will send you a PR for review shortly.

**Comment by Odaeus** [comment]
Fixed in #67.
